File: live_events/CompetitveProgrammingPractice/NOI/2016/DrinkingWater/highprec/drink_sample.pas
// This is an empty program with decimal lib

uses math;

// ---------- decimal lib start ----------

// Decimal_len = PREC div 9 + 1
const Decimal_len : longint = 2100 div 9 + 1;
const Decimal_mo : longint = 1000000000;

type Decimal = record
	is_neg : boolean;
	i : int64;
	data : array[0 .. 2100 div 9] of longint;
end;

function new_decimal() : Decimal;
begin
	new_decimal.is_neg := false;
	new_decimal.i := 0;
	fillchar(new_decimal.data, Decimal_len * 4, 0);
end;

operator := (x : longint) ret : Decimal;
begin
	ret := new_decimal();
	
	if x < 0 then begin
		ret.is_neg := true;
		x := -x;
	end;
	
	ret.i := x;
end;

operator := (x : int64) ret : Decimal;
begin
	ret := new_decimal();
	
	if x < 0 then begin
		ret.is_neg := true;
		x := -x;
	end;
	
	ret.i := x;
end;

operator := (x : double) ret : Decimal;
var
	i : longint;
begin
	ret := new_decimal();
	
	if x < 0 then begin
		ret.is_neg := true;
		x := -x;
	end;
	
	ret.i := floor(x);
	x := x - ret.i;
	
	for i := 0 to Decimal_len - 1 do begin
		x := x * Decimal_mo;
		if x < 0 then x := 0;
		ret.data[i] := floor(x);
		x := x - ret.data[i];
	end;
end;

// Note: ansistring is null-terminated
// see http://wiki.freepascal.org/Character_and_string_types#AnsiString
operator := (s : ansistring) ret : Decimal;
var
	i : longint;
	pos : longint;
	x : longint;
begin
	ret := new_decimal();
	
	// find the first digit or the negative sign
	i := 1;
	while s[i] <> '#0' do begin
		if s[i] = '-' then begin
			ret.is_neg := true;
			inc(i);
			break;
		end else if (s[i] >= '0') and (s[i] <= '9') then begin
			break;
		end;
		inc(i);
	end;
	
	// read the integer part
	while (s[i] >= '0') and (s[i] <= '9') do begin
		ret.i := ret.i * 10 + ord(s[i]) - 48;
		inc(i);
	end;
	
	// read the decimal part
	if s[i] = '.' then begin
		pos := 0;
		x := Decimal_mo div 10;
		
		inc(i);
		while (pos < Decimal_len) and (s[i] >= '0') and (s[i] <= '9') do begin
			ret.data[pos] := ret.data[pos] + (ord(s[i]) - 48) * x;
			inc(i);
			x := x div 10;
			if x = 0 then begin
				inc(pos);
				x := Decimal_mo div 10;
			end;
		end;
	end;
end;

function is_zero(a : Decimal) : boolean;
var
	i : longint;
begin
	if a.i <> 0 then exit(false);
	for i := 0 to Decimal_len - 1 do begin
		if a.data[i] <> 0 then exit(false);
	end;
	exit(true);
end;

function to_double(a : Decimal) : double;
var
	k : double;
	i : longint;
begin
	to_double := a.i;
	
	k := 1.0;
	for i := 0 to Decimal_len - 1 do begin
		k := k / Decimal_mo;
		to_double := to_double + k * a.data[i];
	end;
	
	if a.is_neg then begin
		to_double := -to_double;
	end;
end;

function to_string(a : Decimal; p : longint) : ansistring;
var
	s : ansistring;
	i : longint;
	x, tmp : longint;
begin
	str(a.i, to_string);
	
	if a.is_neg and not is_zero(a) then begin
		to_string := '-' + to_string;
	end;
	
	to_string := to_string + '.';
	
	for i := 0 to Decimal_len - 1 do begin
		// append a.data[i] as "%09d"
		s := '';
		x := Decimal_mo div 10;
		tmp := a.data[i];
		while x <> 0 do begin
			s := s + chr(48 + tmp div x);
			tmp := tmp mod x;
			x := x div 10;
			dec(p);
			if p = 0 then begin
				break;
			end;
		end;
		to_string := to_string + s;
		if p = 0 then break;
	end;
	
	while p > 0 do begin
		to_string := to_string + '0';
		dec(p);
	end;
end;

operator < (a, b : Decimal) ret : boolean;
var
	i : longint;
begin
	if (a.is_neg <> b.is_neg) then begin
		exit(a.is_neg and (not is_zero(a) or not is_zero(b)));
	end else if not a.is_neg then begin
		// a, b >= 0
		if a.i <> b.i then begin
			exit(a.i < b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] < b.data[i]);
			end;
		end;
		exit(false);
	end else begin
		// a, b <= 0
		if a.i <> b.i then begin
			exit(a.i > b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] > b.data[i]);
			end;
		end;
		exit(false);
	end;
end;

operator > (a, b : Decimal) ret : boolean;
var
	i : longint;
begin
	if (a.is_neg <> b.is_neg) then begin
		exit(not a.is_neg and (not is_zero(a) or not is_zero(b)));
	end else if not a.is_neg then begin
		// a, b >= 0
		if a.i <> b.i then begin
			exit(a.i > b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] > b.data[i]);
			end;
		end;
		exit(false);
	end else begin
		// a, b <= 0
		if a.i <> b.i then begin
			exit(a.i < b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] < b.data[i]);
			end;
		end;
		exit(false);
	end;
end;

operator <= (a, b : Decimal) ret : boolean;
var
	i : longint;
begin
	if (a.is_neg <> b.is_neg) then begin
		exit(a.is_neg or (is_zero(a) and is_zero(b)));
	end else if not a.is_neg then begin
		// a, b >= 0
		if a.i <> b.i then begin
			exit(a.i < b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] < b.data[i]);
			end;
		end;
		exit(true);
	end else begin
		// a, b <= 0
		if a.i <> b.i then begin
			exit(a.i > b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] > b.data[i]);
			end;
		end;
		exit(true);
	end;
end;

operator >= (a, b : Decimal) ret : boolean;
var
	i : longint;
begin
	if (a.is_neg <> b.is_neg) then begin
		exit(not a.is_neg or (is_zero(a) and is_zero(b)));
	end else if not a.is_neg then begin
		// a, b >= 0
		if a.i <> b.i then begin
			exit(a.i > b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] > b.data[i]);
			end;
		end;
		exit(true);
	end else begin
		// a, b <= 0
		if a.i <> b.i then begin
			exit(a.i < b.i);
		end;
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> b.data[i] then begin
				exit(a.data[i] < b.data[i]);
			end;
		end;
		exit(true);
	end;
end;

operator = (a, b : Decimal) ret : boolean;
var
	i : longint;
begin
	if is_zero(a) and is_zero(b) then exit(true);
	if a.is_neg <> b.is_neg then exit(false);
	if a.i <> b.i then exit(false);
	for i := 0 to Decimal_len - 1 do begin
		if a.data[i] <> b.data[i] then exit(false);
	end;
	exit(true);
end;

operator <> (a, b : Decimal) ret : boolean;
begin
	exit(not (a = b));
end;

operator - (a : Decimal) ret : Decimal;
var
	i : longint;
begin
	ret := a;
	// -0 = 0
	if not ret.is_neg and (ret.i = 0) then begin
		for i := 0 to Decimal_len - 1 do begin
			if ret.data[i] <> 0 then break;
		end;
		if i < Decimal_len then begin
			ret.is_neg := true;
		end;
	end else begin
		ret.is_neg := not ret.is_neg;
	end;
end;

function _decimal_add(a, b : Decimal) : Decimal;
var
	last : boolean;
	i : longint;
	ret : Decimal;
begin
	ret := a;
	last := false;
	for i := Decimal_len - 1 downto 0 do begin
		ret.data[i] := ret.data[i] + b.data[i] + longint(last);
		if ret.data[i] >= Decimal_mo then begin
			ret.data[i] := ret.data[i] - Decimal_mo;
			last := true;
		end else begin
			last := false;
		end;
	end;
	ret.i := ret.i + b.i + longint(last);
	exit(ret);
end;

operator - (a, b : Decimal) ret : Decimal;
var
	last : boolean;
	i : longint;
begin
	if not a.is_neg and not b.is_neg then begin
		if a >= b then begin
			ret := a;
			last := false;
			for i := Decimal_len - 1 downto 0 do begin
				ret.data[i] := ret.data[i] - b.data[i] - longint(last);
				if ret.data[i] < 0 then begin
					ret.data[i] := ret.data[i] + Decimal_mo;
					last := true;
				end else begin
					last := false;
				end;
			end;
			ret.i := ret.i - b.i - longint(last);
		end else begin
			ret := b;
			last := false;
			for i := Decimal_len - 1 downto 0 do begin
				ret.data[i] := ret.data[i] - a.data[i] - longint(last);
				if ret.data[i] < 0 then begin
					ret.data[i] := ret.data[i] + Decimal_mo;
					last := true;
				end else begin
					last := false;
				end;
			end;
			ret.i := ret.i - a.i - longint(last);
			ret.is_neg := true;
		end;
	end else if a.is_neg and b.is_neg then begin
		// a - b = (-b) - (-a)
		exit(-b - -a);
	end else if a.is_neg then begin
		// -|a| - b
		exit(-_decimal_add(-a, b));
	end else begin
		// a - -|b|
		exit(_decimal_add(a, -b));
	end;
end;

operator + (a, b : Decimal) ret : Decimal;
begin
	if a.is_neg = b.is_neg then begin
		exit(_decimal_add(a, b));
	end else if not a.is_neg then begin
		// a - |b|
		exit(a - -b);
	end else begin
		// b - |a|
		exit(b - -a);
	end;
end;

operator + (a : Decimal; x : int64) ret : Decimal;
var
	last : boolean;
	i : longint;
begin
	if not a.is_neg then begin
		if a.i + x >= 0 then begin
			a.i := a.i + x;
		end else begin
			last := false;
			for i := Decimal_len - 1 downto 0 do begin
				if last or (a.data[i] <> 0) then begin
					a.data[i] := Decimal_mo - a.data[i] - longint(last);
					last := true;
				end else begin
					last := false;
				end;
			end;
			a.i := -x - a.i - longint(last);
			a.is_neg := true;
		end;
	end else begin
		if a.i - x >= 0 then begin
			a.i := a.i - x;
		end else begin
			last := false;
			for i := Decimal_len - 1 downto 0 do begin
				if last or (a.data[i] <> 0) then begin
					a.data[i] := Decimal_mo - a.data[i] - longint(last);
					last := true;
				end else begin
					last := false;
				end;
			end;
			a.i := x - a.i - longint(last);
			a.is_neg := false;
		end;
	end;
	exit(a);
end;

operator + (a : Decimal; x : longint) ret : Decimal;
begin
	ret := a + int64(x);
end;

operator + (x : longint; a : Decimal) ret : Decimal;
begin
	ret := a + int64(x);
end;

operator + (x : int64; a : Decimal) ret : Decimal;
begin
	ret := a + x;
end;

operator + (a : Decimal; x : double) ret : Decimal;
begin
	ret := a + Decimal(x);
end;

operator + (x : double; a : Decimal) ret : Decimal;
begin
	ret := Decimal(x) + a;
end;

operator - (a : Decimal; x : longint) ret : Decimal;
begin
	ret := a + int64(-x);
end;

operator - (x : longint; a : Decimal) ret : Decimal;
begin
	ret := -(a + int64(-x));
end;

operator - (a : Decimal; x : int64) ret : Decimal;
begin
	ret := a + -x;
end;

operator - (x : int64; a : Decimal) ret : Decimal;
begin
	ret := -(a + -x);
end;

operator - (a : Decimal; x : double) ret : Decimal;
begin
	ret := a - Decimal(x);
end;

operator - (x : double; a : Decimal) ret : Decimal;
begin
	ret := Decimal(x) - a;
end;

operator * (a : Decimal; x : longint) ret : Decimal;
var
	last, i : longint;
	tmp : int64;
begin
	if x < 0 then begin
		a.is_neg := not a.is_neg;
		x := -x;
	end else if x = 0 then begin
		exit(new_decimal());
	end;
	
	last := 0;
	for i := Decimal_len - 1 downto 0 do begin
		tmp := int64(a.data[i]) * x + last;
		last := tmp div Decimal_mo;
		a.data[i] := tmp - int64(last) * Decimal_mo;
	end;
	a.i := a.i * x + last;
	
	exit(a);
end;

operator * (x : longint; a : Decimal) ret : Decimal;
begin
	exit(a * x);
end;

operator / (a : Decimal; x : longint) ret : Decimal;
var
	last, i : longint;
	tmp : int64;
begin
	if x < 0 then begin
		a.is_neg := not a.is_neg;
		x := -x;
	end;
	
	last := a.i mod x;
	a.i := a.i div x;
	
	for i := 0 to Decimal_len - 1 do begin
		tmp := int64(last) * Decimal_mo + a.data[i];
		a.data[i] := tmp div x;
		last := tmp - int64(a.data[i]) * x;
	end;
	
	if a.is_neg and (a.i = 0) then begin
		for i := 0 to Decimal_len - 1 do begin
			if a.data[i] <> 0 then begin
				break;
			end;
		end;
		if i = Decimal_len then begin
			a.is_neg := false;
		end;
	end;
	
	exit(a);
end;

// ---------- decimal lib end ----------

begin
	
end.

File: live_events/CompetitveProgrammingPractice/NOI/2016/DrinkingWater/highprec/drink_sample.c
// This is an empty program with decimal lib

#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// ---------- decimal lib start ----------

#define PREC 2100
#define Decimal_len (PREC / 9 + 1)

const int Decimal_mo = 1000000000;

typedef struct {
	bool is_neg;
	long long integer;
	int data[Decimal_len];
} Decimal;

Decimal new_decimal();
Decimal new_decimal_s(const char *s);
Decimal new_decimal_i(int x);
Decimal new_decimal_l(long long x);
Decimal new_decimal_f(double x);

bool is_zero_d(Decimal a);

// p (p > 0) is the number of digits after the decimal point
void to_string_d(Decimal a, char *s, int p);
double to_double_d(Decimal a);

Decimal add_dd(Decimal a, Decimal b);
Decimal add_di(Decimal a, int x);
Decimal add_id(int x, Decimal a);
Decimal add_dl(Decimal a, long long x);
Decimal add_ld(long long x, Decimal a);
Decimal add_df(Decimal a, double x);
Decimal add_fd(double x, Decimal a);

Decimal sub_dd(Decimal a, Decimal b);
Decimal sub_di(Decimal a, int x);
Decimal sub_id(int x, Decimal a);
Decimal sub_dl(Decimal a, long long x);
Decimal sub_ld(long long x, Decimal a);
Decimal sub_df(Decimal a, double x);
Decimal sub_fd(double x, Decimal a);

Decimal mul_di(Decimal a, int x);
Decimal mul_id(int x, Decimal a);

Decimal div_di(Decimal a, int x);

Decimal neg_d(Decimal a);

// <
bool cmp_lt(Decimal a, Decimal b);
// >
bool cmp_gt(Decimal a, Decimal b);
// <=
bool cmp_le(Decimal a, Decimal b);
// >=
bool cmp_ge(Decimal a, Decimal b);
// ==
bool cmp_eq(Decimal a, Decimal b);
// !=
bool cmp_ne(Decimal a, Decimal b);


// implementations

Decimal new_decimal() {
	Decimal ret;
	ret.is_neg = false;
	ret.integer = 0;
	memset(ret.data, 0, Decimal_len * sizeof(int));
	return ret;
}

Decimal new_decimal_s(const char *s) {
	Decimal ret = new_decimal();
	
	// find the first digit or the negative sign
	while (*s != 0) {
		if (*s == '-') {
			ret.is_neg = true;
			++s;
			break;
		} else if (*s >= 48 && *s <= 57) {
			break;
		}
		++s;
	}
	
	// read the integer part
	while (*s >= 48 && *s <= 57) {
		ret.integer = ret.integer * 10 + *s - 48;
		++s;
	}
	
	// read the decimal part
	if (*s == '.') {
		int pos = 0;
		int x = Decimal_mo / 10;
		
		++s;
		while (pos < Decimal_len && *s >= 48 && *s <= 57) {
			ret.data[pos] += (*s - 48) * x;
			++s;
			x /= 10;
			if (x == 0) {
				++pos;
				x = Decimal_mo / 10;
			}
		}
	}
	
	return ret;
}

Decimal new_decimal_i(int x) {
	Decimal ret = new_decimal();
	
	if (x < 0) {
		ret.is_neg = true;
		x = -x;
	}
	
	ret.integer = x;
	
	return ret;
}

Decimal new_decimal_l(long long x) {
	Decimal ret = new_decimal();
	
	if (x < 0) {
		ret.is_neg = true;
		x = -x;
	}
	
	ret.integer = x;
	
	return ret;
}

Decimal new_decimal_f(double x) {
	Decimal ret = new_decimal();
	
	if (x < 0) {
		ret.is_neg = true;
		x = -x;
	}
	
	ret.integer = (long long)x;
	x -= ret.integer;
	
	int i;
	for (i = 0; i < Decimal_len; i++) {
		x *= Decimal_mo;
		if (x < 0) x = 0;
		ret.data[i] = (int)x;
		x -= ret.data[i];
	}
	
	return ret;
}

bool is_zero_d(Decimal a) {
	if (a.integer) return false;
	int i;
	for (i = 0; i < Decimal_len; i++) {
		if (a.data[i]) return false;
	}
	return true;
}

void to_string_d(Decimal a, char *s, int p) {
	if (a.is_neg && !is_zero_d(a)) {
		*s++ = '-';
	}
	
	// append the integer part to s
	long long x = a.integer;
	if (x == 0) {
		*s++ = 48;
	} else {
		char _[30];
		int cnt = 0;
		while (x) {
			_[cnt++] = x % 10;
			x /= 10;
		}
		while (cnt--) {
			*s++ = _[cnt] + 48;
		}
	}
	
	// append the decimal part to s
	*s++ = '.';
	int i;
	for (i = 0; i < Decimal_len; i++) {
		// append a.data[i] as "%09d"
		int x = Decimal_mo / 10;
		int tmp = a.data[i];
		while (x) {
			*s++ = 48 + tmp / x;
			tmp %= x;
			x /= 10;
			if (--p == 0) {
				break;
			}
		}
		if (p == 0) break;
	}
	if (p > 0) {
		while (p--) {
			*s++ = '0';
		}
	}
	
	*s++ = '\0';
}

double to_double_d(Decimal a) {
	double ret = a.integer;
	
	double k = 1.0;
	int i;
	for (i = 0; i < Decimal_len; i++) {
		k /= Decimal_mo;
		ret += k * a.data[i];
	}
	
	if (a.is_neg) {
		ret = -ret;
	}
	
	return ret;
}

bool cmp_lt(Decimal a, Decimal b) {
	if (a.is_neg != b.is_neg) {
		return a.is_neg && (!is_zero_d(a) || !is_zero_d(b));
	} else if (!a.is_neg) {
		// a, b >= 0
		if (a.integer != b.integer) {
			return a.integer < b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] < b.data[i];
			}
		}
		return false;
	} else {
		// a, b <= 0
		if (a.integer != b.integer) {
			return a.integer > b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] > b.data[i];
			}
		}
		return false;
	}
}

bool cmp_gt(Decimal a, Decimal b) {
	if (a.is_neg != b.is_neg) {
		return !a.is_neg && (!is_zero_d(a) || !is_zero_d(b));
	} else if (!a.is_neg) {
		// a, b >= 0
		if (a.integer != b.integer) {
			return a.integer > b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] > b.data[i];
			}
		}
		return false;
	} else {
		// a, b <= 0
		if (a.integer != b.integer) {
			return a.integer < b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] < b.data[i];
			}
		}
		return false;
	}
}

bool cmp_le(Decimal a, Decimal b) {
	if (a.is_neg != b.is_neg) {
		return a.is_neg || (is_zero_d(a) && is_zero_d(b));
	} else if (!a.is_neg) {
		// a, b >= 0
		if (a.integer != b.integer) {
			return a.integer < b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] < b.data[i];
			}
		}
		return true;
	} else {
		// a, b <= 0
		if (a.integer != b.integer) {
			return a.integer > b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] > b.data[i];
			}
		}
		return true;
	}
}

bool cmp_ge(Decimal a, Decimal b) {
	if (a.is_neg != b.is_neg) {
		return !a.is_neg || (is_zero_d(a) && is_zero_d(b));
	} else if (!a.is_neg) {
		// a, b >= 0
		if (a.integer != b.integer) {
			return a.integer > b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] > b.data[i];
			}
		}
		return true;
	} else {
		// a, b <= 0
		if (a.integer != b.integer) {
			return a.integer < b.integer;
		}
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != b.data[i]) {
				return a.data[i] < b.data[i];
			}
		}
		return true;
	}
}

bool cmp_eq(Decimal a, Decimal b) {
	if (is_zero_d(a) && is_zero_d(b)) return true;
	if (a.is_neg != b.is_neg) return false;
	if (a.integer != b.integer) return false;
	int i;
	for (i = 0; i < Decimal_len; i++) {
		if (a.data[i] != b.data[i]) return false;
	}
	return true;
}

bool cmp_ne(Decimal a, Decimal b) {
	return !cmp_eq(a, b);
}

Decimal neg_d(Decimal a) {
	Decimal ret = a;
	// -0 = 0
	if (!ret.is_neg && ret.integer == 0) {
		int i;
		for (i = 0; i < Decimal_len; i++) {
			if (ret.data[i] != 0) break;
		}
		if (i < Decimal_len) {
			ret.is_neg = true;
		}
	} else {
		ret.is_neg ^= 1;
	}
	return ret;
}

Decimal add_dd(Decimal a, Decimal b) {
	if (a.is_neg == b.is_neg) {
		Decimal ret = a;
		bool last = true;
		int i;
		for (i = Decimal_len - 1; i >= 0; i--) {
			ret.data[i] += b.data[i] + last;
			if (ret.data[i] >= Decimal_mo) {
				ret.data[i] -= Decimal_mo;
				last = true;
			} else {
				last = false;
			}
		}
		ret.integer += b.integer + last;
		return ret;
	} else if (!a.is_neg) {
		// a - |b|
		return sub_dd(a, neg_d(b));
	} else {
		// b - |a|
		return sub_dd(b, neg_d(a));
	}
}

Decimal sub_dd(Decimal a, Decimal b) {
	if (!a.is_neg && !b.is_neg) {
		if (cmp_ge(a, b)) {
			Decimal ret = a;
			bool last = false;
			int i;
			for (i = Decimal_len - 1; i >= 0; i--) {
				ret.data[i] -= b.data[i] + last;
				if (ret.data[i] < 0) {
					ret.data[i] += Decimal_mo;
					last = true;
				} else {
					last = false;
				}
			}
			ret.integer -= b.integer + last;
			return ret;
		} else {
			Decimal ret = b;
			bool last = false;
			int i;
			for (i = Decimal_len - 1; i >= 0; i--) {
				ret.data[i] -= a.data[i] + last;
				if (ret.data[i] < 0) {
					ret.data[i] += Decimal_mo;
					last = true;
				} else {
					last = false;
				}
			}
			ret.integer -= a.integer + last;
			ret.is_neg = true;
			return ret;
		}
	} else if (a.is_neg && b.is_neg) {
		// a - b = (-b) - (-a)
		return sub_dd(neg_d(b), neg_d(a));
	} else if (a.is_neg) {
		// -|a| - b
		return neg_d(add_dd(neg_d(a), b));
	} else {
		// a - -|b|
		return add_dd(a, neg_d(b));
	}
}

Decimal add_dl(Decimal a, long long x) {
	if (!a.is_neg) {
		if (a.integer + x >= 0) {
			a.integer += x;
		} else {
			bool last = false;
			int i;
			for (i = Decimal_len - 1; i >= 0; i--) {
				if (last || a.data[i]) {
					a.data[i] = Decimal_mo - a.data[i] - last;
					last = true;
				} else {
					last = false;
				}
			}
			a.integer = -x - a.integer - last;
			a.is_neg = true;
		}
	} else {
		if (a.integer - x >= 0) {
			a.integer -= x;
		} else {
			bool last = false;
			int i;
			for (i = Decimal_len - 1; i >= 0; i--) {
				if (last || a.data[i]) {
					a.data[i] = Decimal_mo - a.data[i] - last;
					last = true;
				} else {
					last = false;
				}
			}
			a.integer = x - a.integer - last;
			a.is_neg = false;
		}
	}
	return a;
}

Decimal add_ld(long long x, Decimal a) {
	return add_dl(a, x);
}

Decimal add_di(Decimal a, int x) {
	return add_dl(a, x);
}

Decimal add_id(int x, Decimal a) {
	return add_dl(a, x);
}

Decimal sub_dl(Decimal a, long long x) {
	return add_dl(a, -x);
}

Decimal sub_ld(long long x, Decimal a) {
	return neg_d(add_dl(a, -x));
}

Decimal sub_di(Decimal a, int x) {
	return add_dl(a, -x);
}

Decimal sub_id(int x, Decimal a) {
	return neg_d(add_dl(a, -x));
}

Decimal add_df(Decimal a, double x) {
	return add_dd(a, new_decimal_f(x));
}

Decimal add_fd(double x, Decimal a) {
	return add_dd(new_decimal_f(x), a);
}

Decimal sub_df(Decimal a, double x) {
	return add_dd(a, new_decimal_f(-x));
}

Decimal sub_fd(double x, Decimal a) {
	return neg_d(add_dd(a, new_decimal_f(-x)));
}

Decimal mul_di(Decimal a, int x) {
	if (x < 0) {
		a.is_neg ^= 1;
		x = -x;
	} else if (x == 0) {
		return new_decimal();
	}
	
	int last = 0;
	int i;
	for (i = Decimal_len - 1; i >= 0; i--) {
		long long tmp = 1LL * a.data[i] * x + last;
		last = tmp / Decimal_mo;
		a.data[i] = tmp - 1LL * last * Decimal_mo;
	}
	a.integer = a.integer * x + last;
	
	return a;
}

Decimal mul_id(int x, Decimal a) {
	return mul_di(a, x);
}

Decimal div_di(Decimal a, int x) {
	if (x < 0) {
		a.is_neg ^= 1;
		x = -x;
	}
	
	int last = a.integer % x;
	a.integer /= x;
	
	int i;
	for (i = 0; i < Decimal_len; i++) {
		long long tmp = 1LL * last * Decimal_mo + a.data[i];
		a.data[i] = tmp / x;
		last = tmp - 1LL * a.data[i] * x;
	}
	
	if (a.is_neg && a.integer == 0) {
		for (i = 0; i < Decimal_len; i++) {
			if (a.data[i] != 0) {
				break;
			}
		}
		if (i == Decimal_len) {
			a.is_neg = false;
		}
	}
	
	return a;
}

// ---------- decimal lib end ----------

int main() {
	return 0;
}
